File: lintcode/cpp/Merge_k_Sorted_Lists.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */

/** Time complexity: 
 *  1. construct the priority_queue of size k. lg1 + lg2 + lg3 + lgk = lg(k!)
 *  2. merge the rest of the node, and each one has complexity O(N*lgk), where N is the rest of the nodes.
 */
struct listNode_cmp{
    bool operator()(ListNode *lhs, ListNode *rhs){
        return lhs->val > rhs->val;
    }
};

class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // write your code here
        priority_queue<ListNode *, vector<ListNode *>, listNode_cmp> minHeap;
        
        for(int i = 0; i < lists.size(); i++){
            if(lists[i]) minHeap.push(lists[i]);
        }
        
        ListNode dummy(0);
        ListNode *d = &dummy;
        
        while(!minHeap.empty()){
            ListNode *t = minHeap.top();
            d->next = t;
            d = d->next;
            minHeap.pop();
            if(t->next) minHeap.push(t->next);
        }
        
        return dummy.next;
    }
};

File: lintcode/java/MergekSortedLists.java
/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    public ListNode mergeKLists(List<ListNode> lists) {  
        if(lists == null) throw new IllegalArgumentException("xxx");
        if(lists.isEmpty()) return null;
        
        // can be replaced with lambda functions
        Queue<ListNode> pQueue = new PriorityQueue<ListNode>(lists.size(), new Comparator<ListNode>() {
            @Override
            public int compare(ListNode lhs, ListNode rhs) {
                if(lhs.val > rhs.val) return 1;
                else if(lhs.val < rhs.val) return -1;
                else return 0;
            }
        });
        
        for(ListNode list: lists){
            if(list != null) pQueue.offer(list);
        }
        
        ListNode dummy = new ListNode(0);
        ListNode d = dummy;
        while(!pQueue.isEmpty()){
            ListNode t = pQueue.poll();
            d.next = t;
            d = d.next;
            if(t.next != null) pQueue.offer(t.next);
        }
        
        return dummy.next;
    }
}
